First draft of advanced section for iterstreams (#55)
merantix-momentum/squirrel-core: docs/advanced/iterstream.rst

IterStream
==========

> The IterStream module provides functionalities to chain operations on iterables. This functionality

Reviewer: ```suggestion
The IterStream module provides functionalities to chain iterables. This functionality
```

> is provided through the mix-in class :py:class:`Composable`, which implements methods such as map or filter.

Reviewer: Composable is not mix-in

> The nice thing about it is that all operations are done lazily, that is no computation is executed until we ask for it.

Reviewer: more clear wording here, maybe say squirrel pipeline is lazy. Then before hand, the term `pipeline` should be defined, probably best explained by an example

Author: speaking of which, what is a `pipeline` ? Is it simply a chain of Composables ?

.. code-block:: python

    from squirrel.iterstream import IterableSource

    def add_1(x):
        print(f'add 1 to {x}')
        return x + 1

    def mult_10(x):
        print(f'multiply 10 to {x}')
        return x * 10

    it = IterableSource(range(3)).map(add_1).map(mult_10)
    next(iter(it))

Output::

    add 1 to 0
    multiply 10 to 1

Here, we see that the functions are only executed when the iterator is asked to fetch the next item.

> An alternative way of chaining operations is via :py:meth:`squirrel.iterstream.base.Composable.compose`.

Reviewer: please highlight this, maybe by adding a dedicated sub-section, that this is the way to customize and extend iterstream api, and that it is meant to be used when the standard Composable methods are insufficient to express the logic. 

please also mention that if pipeline should be passed to tut.DataLoader, is must be picklable, and if the pipeline contains a non-picklable object such as an instance of a fsspec filesyste, the instantiation should happen in the iter method.

.. code-block:: python

    from squirrel.iterstream import IterableSource
    from squirrel.iterstream import Composable

    class Add_1(Composable):
    def __init__(self):
        pass

    def __iter__(self):
        for x in self.source:
            print(f'add 1 to {x}')
            yield x + 1

    class Mult_10(Composable):
        def __init__(self):
            pass

        def __iter__(self):
            for x in self.source:
                print(f'multiply 10 to {x}')
                yield x * 10

    it = IterableSource(range(3)).compose(Add_1).compose(Mult_10)
    next(iter(it))

Output::

    add 1 to 0
    multiply 10 to 1

Similar as before, the execution is done lazily. The only difference is that we wrap the function inside a class inheriting
from `Composable`. One scenario why we might want to use `compose()` is when we need to instantiate an expensive resource

> for the transformations such as a large network for preprocessing or a database connection.

Reviewer: what does `large network for preprocessing` mean?

Author: extract word embeddings or image features from a large model like bert

An interesting fact about `compose()` is its order of execution:

.. code-block:: python

    from squirrel.iterstream import IterableSource
    from squirrel.iterstream import Composable

    class Add_1(Composable):
    def __init__(self):
        pass

    def __iter__(self):
        for x in self.source:
            print(f'add 1 to {x}')
            yield x + 1

    class Mult_10(Composable):
        def __init__(self):
            pass

        def __iter__(self):
            for x in self.source:
                print(f'multiply 10 to {x}')
                yield x * 10

    it = IterableSource(range(3)).compose(Add_1).compose(Mult_10)
    next(iter(it))

Output::

>     Create Add_1

Reviewer: the print statement for this is missing in the code?

    Create Mult_10
    Start Mult_10
    Start Add_1
    add 1 to 0
    multiply 10 to 1

The constructors are called from left to right, as is the execution of the transformations. However, we can see
that the iterators are called from right to left.

Architecture
--------------------

> Most classes inherit from `Composable` to implement methods for stream manipulation.

Reviewer: mention that these are internal classes not meant to be used by users

:py:meth:`squirrel.iterstream.base._Iterable` and :py:meth:`squirrel.iterstream.base._AsyncMap` are both instantiated when

> chaining stream manipulations. The difference is that `_AyncMap` deals with asynchronous execution. We show a non-exhaustive UML

Reviewer: a bit more on what `_AsyncMap` does is helpful. 

Please mention that the order is preserved with `_AsyncMap`, maybe worth mentioning in the basic usage. Here we can explain the internal queue of size `buffer` that contains `AsyncContent` objects

       class _Iterable {
            source: Iterable
            callback: Callable

            __iter__() Iterator
       }

       class IterableSource {
            source: Iterable or Callable

            __iter__() Iterator
       }

        class _AsyncMap {
            source: Iterable
            callback: Callable
            buffer: int
            max_workers: int
            executor: Executor

            __iter__() Iterator
       }

       class AsyncContent {
            future: executor.submit(func, item)

            value(): fetch results
       }

Special Composables
---------------------
There are a few special `Composable` classes such as :py:class:`squirrel.iterstream.source.FilePathGenerator` or
classes that help you interface with PyTorch. `FilePathGenerator` for example is an iterable that returns a generator iterator
over the folder contents given an url.
We can also transform an iterable into a PyTorch compatible format using
:py:class:`squirrel.iterstream.torch_composables.TorchIterable`.
The tutorial on
`PyTorch training with Squirrel <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/03.Pytorch_Model_Training.ipynb/>`_

> provides a good example usage. Other special `Composable` classes such as :py:class:`squirrel.iterstream.torch_composables.SplitByWorker`

Reviewer: ```suggection
provides an example. Other special `Composable` classes such as :py:class:`squirrel.iterstream.torch_composables.SplitByWorker`
```

IterStream
==========
The IterStream module provides functionalities to chain iterables. This functionality
is provided through the :py:class:`Composable` class, which forms the base class for most classes in IterStream.

> 

Author: hmm what do we want to show here ? This chain of generators has the same lazy execution property as composables, right?

    can't have a non-pickable object such as a `fssepc` object. A solution is to create the object in the `__iter__` method
    instead of inside the constructor.

Asynchronous Execution
-----------------------
The documentation in :ref:`usage/iterstream:IterStream` explain how asynchronous execution is performed with :py:meth:`async_map`.
Internally, a :py:class:`_AsyncMap` object is constructed when calling :py:meth:`async_map`.
:py:class:`_AsyncMap` maintains an internal queue and creates :py:class:`AsyncContent` that are inserted to the queue.
:py:class:`AsyncContent` objects are created by specifying a function callback, the item it operates on and an executor.
When :py:class:`AsyncContent` object is created, the function callback is scheduled for asynchronous execution. We can simply fetch results
from the queue by iterating over the :py:class:`_AsyncMap` object.

Stream Processing Methods
--------------------
The :py:class:`Composable` class offers roughly three groups of methods for processing streams.

* *Transformations* : :py:meth:`map`, :py:meth:`filter`. These methods can be used to apply a transformation over the stream.

> * *Terminal* : :py:meth:`join`, py:meth:`collect`. These methods are used to materialize the stream.

Reviewer: Materialize it not technically correct, as items in the stream are already in memory and materialized. Terminal operations in fact consume the stream and turn in into a non-composable object, e.g. None and List.

Author: so the whole iterator is loaded in memory ? what is the benefit of lazy execution then ?

> * *Organization*: :py:meth:`shuffle`, py:meth:`batch`, py:meth:`take`. THe methods are used to order and organize the stream.

Reviewer: shuffle and batch are part of transformation.
take is a terminal operation
Instead of `Organization`, there should be another kind `Source`, should be mentioned as the first class I guess, and includes e.g. `IterableSource`.

Author: what's the function of source ?

IterStream
==========
The IterStream module provides functionalities to chain iterables. The core principle behind IterStream is
similar to chaining generators, that is lazy execution.

.. code-block:: python

>     from squirrel.iterstream import IterableSource

Reviewer: this import statement is not used here, please move to the next code block where it's used

    def add_1(x):
        print(f'add 1 to {x}')
        return x + 1

    def mult_10(x):
        print(f'multiply 10 to {x}')
        return x * 10

    g1 = (add_1(i) for i in range(3))
    g2 = (mult_10(i) for i in g1)
    next(iter(g2))

Output::

    add 1 to 0
    multiply 10 to 1

This functionality is provided through the :py:class:`Composable` class, which forms the base class for most classes in IterStream.

.. code-block:: python

    it = IterableSource(range(3)).map(add_1).map(mult_10)
    next(iter(it))

Output::

    add 1 to 0
    multiply 10 to 1

> In the example above we see how Composables are chained. We also call this chain of Composables **stream**. The executions are

Reviewer: at this point it's still not shown that `it` is composable. I suggest the following:

```
assert isinstance(IterableSource(range(3)), Composable)
assert isinstance(IterableSource(range(3).map(add_1)), Composable)
assert isinstance(IterableSource(range(3).map(add_1).map(mult_10)), Composable)
```

Author: It is a bit verbose though. Is it maybe enough to describe that IterableSource inherits from Composable ?

done lazily, that is the transformation is only executed when the iterator fetches the next item.

Custom Composable
--------------------
An alternative way of constructing streams is via :py:meth:`squirrel.iterstream.base.Composable.compose`.

.. code-block:: python

    from squirrel.iterstream import IterableSource
    from squirrel.iterstream import Composable

    class Add_1(Composable):
    def __init__(self):
        pass

    def __iter__(self):
        for x in self.source:
            print(f'add 1 to {x}')

>             yield x + 1

Reviewer: wrong indentation

    class Mult_10(Composable):
        def __init__(self):
            pass

        def __iter__(self):
            for x in self.source:
                print(f'multiply 10 to {x}')
                yield x * 10

    it = IterableSource(range(3)).compose(Add_1).compose(Mult_10)
    next(iter(it))

Output::

    add 1 to 0
    multiply 10 to 1

Similar as before, the execution is done lazily. The only difference is that we wrap the function inside a custom :py:class:`Composable`
class. Writing custom Composable classes allows us to modify the iteration process. Some use-cases for custom Composables include:

>     #. We need to instantiate a expensive resource only once in the constructor e.g. a database connection or a R-CNN feature extractor.

Reviewer: ```suggestion
    #. We need to instantiate an expensive resource such as a database connection or a R-CNN feature extractor only once for the entire stream instead of once per item. In this case it is advisable to instantiate this resource in the `__iter__` method of your custom class. One advantage is that the object is only instantiated once iterating over the stream starts. Additionally, if this resource is a non-picklable object, the stream can still be pickled before the iteration starts (e.g. in the multiprocessing context). 

```

    #. When a very complex  stream processing is needed that is hard to achieve with standard methods.

> When using `compose()` note that the order of calling the `__iter__` method is from right ot left.

Reviewer: Not only the case for `compose()`, but for all other methods too, e.g. the same is the case when you have `map()`, but it's just easier to illustrate the concept with print statements with `compose`

> When using `compose()` note that the order of calling the `__iter__` method is from right ot left.

Reviewer: maybe a more general statement, that the following example illustrates the control flow and order of instantiation and iterations of a composable object that comprises several transformations.

        def __init__(self):
            print("Create Add_1")
            super().__init__()

        def __iter__(self):
            print("Start Add_1")
            for i in iter(self.source):
                print(f"add 1 to {i+1}")
                yield i+1

    class Mult_10(Composable):
        def __init__(self):
            print("Create Mult_10")
            super().__init__()

        def __iter__(self):
            print("Start Mult_10")
            for i in iter(self.source):
                print(f"multiply 10 to {10*i}")
                yield 10*i

    it = IterableSource(range(3)).compose(Add_1).compose(Mult_10)
    next(iter(it))

Output::

    Create Add_1
    Create Mult_10
    Start Mult_10
    Start Add_1
    add 1 to 0
    multiply 10 to 1

The constructors are called from left to right, as is the execution of the transformations. However, we can see
that the iterators are called from right to left.

.. admonition:: PyTorch

    There are already special Composables implemented for interfacing with PyTorch such as :py:class:`TorchIterable` or
    :py:class:`SplitByWorker`. Examples are given in the :ref:`PyTorch integration section <pytorch_example>`.
..

>     Note that PyTorch Dataloader requires the iterable passed to be pickable when using multi-processing. That is, our custom Composable

Reviewer: probably better to mention this in the PyTorch integration section

Author: This is already explained in the PyTorch integration section, I thought of it more like a link/hint. Maybe it's redundant though, we can just remove it.

Output::

    x: [0, 1, 2]
    x + 1: [1, 2, 3]
    (x + 1) * 10: [10, 20, 30]

Asynchronous execution
----------------------
We have seen in :ref:`usage/iterstream` how to apply functions on streams asynchronously with :py:meth:`async_map`.
Internally, a :py:class:`_AsyncMap` object is constructed when calling :py:meth:`async_map`.
:py:class:`_AsyncMap` maintains an internal queue and creates :py:class:`AsyncContent` that are inserted to the queue.
:py:class:`AsyncContent` objects are created by specifying a function callback, the item it operates on, and an executor.
When :py:class:`AsyncContent` object is created, the function callback is scheduled for asynchronous execution. We can simply fetch results
from the queue by iterating over the :py:class:`_AsyncMap` object.

Architecture
--------------------
Most classes inherit from :py:class:`Composable`, which provides many concrete stream processing methods such as :py:meth:`map` and :py:meth:`filter`, and one abstract method `__iter__` which must be implemented by all sub classes. 

In the non-exhaustive UML diagram below, we show how the IterStream module is structured.

.. mermaid::

>     classDiagram

Reviewer: please signal that Composable is abstract

Internally, a :py:class:`_AsyncMap` object is constructed when calling :py:meth:`async_map`.
:py:class:`_AsyncMap` maintains an internal queue and creates :py:class:`AsyncContent` that are inserted to the queue.
:py:class:`AsyncContent` objects are created by specifying a function callback, the item it operates on, and an executor.
When :py:class:`AsyncContent` object is created, the function callback is scheduled for asynchronous execution. We can simply fetch results
from the queue by iterating over the :py:class:`_AsyncMap` object.

Architecture
--------------------
Most classes inherit from :py:class:`Composable` to implement methods for stream processing.
In the non-exhaustive UML diagram below, we show how the IterStream module is structured.

.. mermaid::

    classDiagram

        Composable <|-- _Iterable
        Composable <|-- IterableSource
        Composable <|-- _AsyncMap
        AsyncContent <.. _AsyncMap

        class Composable {
            source: Iterable or Callable

            __iter__() Iterator
            compose(constructor, *args, **kwargs) Composable
            map(callback) _Iterable

>             async_map(callback, buffer, max_workers, executor) _Iterable

Reviewer: ```suggestion
            async_map(callback, buffer, max_workers, executor) _AsyncMap
```

IterStream
==========

> 

Reviewer: Open the section with a sentence about what iterstream api aims to do: chaining iterators. The the example of chaining generators, then provide the equivalent implementation with `.compose(MyIterable)`, then explain the control flow (maybe with some print statements in the init and iter methods). Then explain that public methods on Composable such as map and filter are using internal classes for wrapping user defined function. After this the UML diagram will make more sense.

Author: - It seems like we are making a difference beween chaining iterators and chaining generators. Is there a difference?
- What is `.compose(MyIterable)` equivalent to ? My understanding is that it constructs a new Composable with MyIterable. 
- What should the code examples show here ? That we can chain operations ? That the order is reversed ?

Architecture
--------------------
To see how the IterStream module is structured, let's see the UML diagram. This not an exhaustive diagram and only
aims to show the most important concepts through the class relations.

> .. mermaid::

Author: Not sure if this diagram in itself is so telling

Reviewer: Please include `AsyncContent` and its relationship with `_AsyncMap`

    classDiagram

        Composable <|-- _Iterable
        Composable <|-- _AsyncMap
        Composable <|-- IterableSource

        class Composable {
            source: Iterable or Callable

            __iter__() Iterator
            map(callback) _Iterable
            async_map(callback, buffer, max_workers, executor) _Iterable
            to(callback, *args, **kwargs) _Iterable
            compose(constructor, *args, **kwargs) Composable
        }

       class _Iterable {
            source: Iterable
            callback: Callable
            buffer: int
            max_workers: int
            executor: Executor

            __iter__() Iterator
       }

       class _AsyncMap {
            source: Iterable

>             callback: Callable

Reviewer: ```suggestion
            callback: Callable
            int buffer
            int max_workers
            Executor executor
```

            __iter__() Iterator
       }

       class IterableSource {
            source: Iterable or Callable

            __iter__() Iterator
       }

Composable Class
---------------------
:py:class:`Composable` is a mix-in class that implements utility methods to chain operations on streams of data.
Each :py:class:`Composable` contains a source Iterable on which the methods operate.
The most prominent ones been :py:meth:`squirrel.iterstream.base.Composable.async_map` and :py:meth:`squirrel.iterstream.base.Composable.map` that
allow for transformations over the whole iterable. Other utility methods include

> * :py:meth:`squirrel.iterstream.base.Composable.batched`: explaination

Author: Listing all functions is maybe a bit too verbose ?

Reviewer: agree

* :py:meth:`squirrel.iterstream.base.Composable.shuffle`: ..
* :py:meth:`squirrel.iterstream.base.Composable.take`
* :py:meth:`squirrel.iterstream.base.Composable.filter`
* :py:meth:`squirrel.iterstream.base.Composable.join`
* :py:meth:`squirrel.iterstream.base.Composable.collect`
* :py:meth:`squirrel.iterstream.base.Composable.flatten`
* :py:meth:`squirrel.iterstream.base.Composable.to`
* :py:meth:`squirrel.iterstream.base.Composable.compose`

> .. admonition:: to vs compose

Reviewer: I suggest to make this a dedicated section about extending iterStream with user defined iterables.

    Both :py:meth:`squirrel.iterstream.base.Composable.to` and :py:meth:`squirrel.iterstream.base.Composable.compose`
    instantiate and return new :py:class:`Composable` objects from the source Iterable. In the latter case any class inheriting
    from :py:class:`Composable` can be used. This can be for example an :py:class`TorchIterable` that
    allows for interfacing with the PyTorch DataLoader (`model training with PyTorch  <https://github.com/merantix-momentum/squirrel-datasets-core/blob/main/examples/03.Pytorch_Model_Training.ipynb/>`_).
    For the former case, :py:class:`_Iterable` object is constructed using the source and a callback function.
    This class is an Iterable that returns an Iterator where the callback function is applied over the source.

We have already came across the :py:class:`_Iterable` class when talking about :py:meth:`squirrel.iterstream.base.Composable.to`.

> A very similar class is the :py:class:`__AsyncMap` class ...

Author: So ` _AsyncMap` is mostly used with `async_map ` method, but ` _Iterable` is also used with `map` (but other methods also return ` _Iterable`).  Is this a distinction we want to mention here?

    There are already special Composables implemented for interfacing with PyTorch such as :py:class:`TorchIterable` or
    :py:class:`SplitByWorker`. Examples are given in :ref:`usage/iterstream:PyTorch Distributed Dataloading`.
..
    Note that PyTorch Dataloader requires the iterable passed to be pickable when using multi-processing. That is, our custom Composable
    can't have a non-pickable object such as a `fssepc` object. A solution is to create the object in the `__iter__` method
    instead of inside the constructor.

Stream Processing Methods
-------------------------
The :py:class:`Composable` class offers three kinds of methods for processing streams.

> * *Source*: :py:meth:`source_``. Source defines the source iterable on which transformations are applied.

Reviewer: ```suggestion
* *Source*: The first node in the stream that generates items or wraps an iterable, for instance :py:class:`IterableSource`.

```

> * *Transformations* : :py:meth:`map`, :py:meth:`filter`, :py:meth:`shuffle`, :py:meth:`batch`. These methods can be used to apply a transformation over the stream.

Reviewer: ```suggestion
* *Transformations* : Provide a way to apply transformations on items in the stream, such as :py:meth:`map` and :py:meth:`filter`, or manipulate the stream itself, such as :py:meth:`shuffle`, :py:meth:`batched`. 
```

* *Terminal* : :py:meth:`join`, py:meth:`collect`. These methods are used to consume the stream.

Source in a Stream
------------------------
In a stream, each `Composable` in the chain stores the iterable it operates on in the `source` attribute. That is if we
get the `source` from the *n*-th `Composable` in the chain, we can retrieve the intermediate
results up until the *n-1*-th `Composable` (including). However, note that after repeatedly calling
`source` we will end up with the original iterable, which will not have a `source` attribute.

.. code-block:: python

    def add_1(x):
        return x + 1

    def mult_10(x):
        return x * 10

    it = IterableSource(range(3)).map(add_1).map(mult_10)

    print(f'x: {it.source.source.collect()}')
    print(f'x + 1: {it.source.collect()}')
    print(f'(x + 1) * 10: {it.collect()}')

Output::

    x: [0, 1, 2]
    x + 1: [1, 2, 3]

>     (x + 1) * 10: [10, 20, 30]

Reviewer: :+1:
